### PR TITLE
Mint capability added + Capabilities refactored

### PR DESCRIPTION
**Pull request by mjgrzybek**
1. Mint capability added using copy-paste pattern
2. Capabilities part refactored:
  * `capability` package renamed to `activegate`
  * `capability` used as `activegate` specialization
  * `capability.Capability` created with `MakeCapability`

### CODE REVIEW

File: config/crd/bases/dynatrace.com_dynakubes.yaml [hunk 1]
```diff
@@ -2117,6 +2117,211 @@ spec:
                     type: object
                   type: array
               type: object
+            metricsIngest:
+              description: ' Configuration for Metrics Ingest'
```
Comment (DTMad) [inline config/crd/bases/dynatrace.com_dynakubes.yaml:2121]: Space before first letter on purpose?

File: controllers/activegate/capability/reconciler.go [hunk 2]
```diff
@@ -63,36 +72,42 @@ func setCommunicationsPort(_ *dynatracev1alpha1.DynaKube) capability.StatefulSet
...
 }
 
 func (r *Reconciler) createServiceIfNotExists() (bool, error) {
-	service := createService(r.Instance, Module)
+	service := createService(r.Instance, r.ModuleName)
 
 	err := r.Get(context.TODO(), client.ObjectKey{Name: service.Name, Namespace: service.Namespace}, service)
 	if err != nil && k8serrors.IsNotFound(err) {
-		r.log.Info("creating service for router")
+		r.log.Info("creating service for " + r.ModuleName)
```
Comment (DTMad) [inline controllers/activegate/capability/reconciler.go:110]: You can make use of the key - value list of the logging function here 
e.g. r.log.Info("msg", "key", someValue)

**Review by DTMad — CHANGES_REQUESTED** [review]

File: config/common/kustomization.yaml [hunk 3]
```diff
@@ -15,6 +15,9 @@ resources:
   - routing/serviceaccount-routing.yaml
   - routing/role-routing.yaml
   - routing/rolebinding-routing.yaml
+  - mint/serviceaccount-mint.yaml
+  - mint/role-mint.yaml
+  - mint/rolebinding-mint.yaml
```
Comment (baichinger) [inline config/common/kustomization.yaml:20]: s/mint/metrics/

File: config/kubernetes/kustomization.yaml [hunk 4]
```diff
@@ -4,6 +4,7 @@ resources:
   - operator/podsecuritypolicy-operator.yaml
   - kubernetes-monitoring/podsecuritypolicy-kubernetes-monitoring.yaml
   - routing/podsecuritypolicy-routing.yaml
+  - mint/podsecuritypolicy-mint.yaml
```
Comment (baichinger) [inline config/kubernetes/kustomization.yaml:7]: s/mint/metrics/

File: controllers/activegate/capability/capability.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+func (c *Capability) CalculateStatefulSetName(instanceName string) string {
+	return instanceName + "-" + c.ModuleName
+}
+
+type CapabilityType int
+
+const (
+	Kubemon CapabilityType = iota
+	Routing
+	Mint
```
Comment (baichinger) [inline controllers/activegate/capability/capability.go:31]: s/Mint/Metrics/

File: config/crd/bases/dynatrace.com_dynakubes.yaml (hunk 1 above)
Comment (mjgrzybek) [reply config/crd/bases/dynatrace.com_dynakubes.yaml:2121]: Yes, it follows Kubemon and Routing convention.

File: controllers/dynakube/dynakube_controller.go [hunk 6]
```diff
@@ -235,35 +220,50 @@ func (r *ReconcileDynaKube) ensureDeleted(obj client.Object) error {
...
-		sts := appsv1.StatefulSet{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      rec.Instance.Name + routing.StatefulSetSuffix,
-				Namespace: rec.Instance.Namespace,
-			},
-		}
-		if err := r.ensureDeleted(&sts); rec.Error(err) {
-			return false
-		}
+func (r *ReconcileDynaKube) reconcileCaps(rec *utils.Reconciliation, dtc dtclient.Client) bool {
```
Comment (baichinger) [inline controllers/dynakube/dynakube_controller.go:223]: This method takes care to reconcile ActiveGate capabilities. Therefore suggest to use `reconcileActiveGateCapabilities`.

**Comment by baichinger** [comment]
Suggest to push those changes to upstream and create a PR there. This will make testing this out much easier.

File: controllers/activegate/capability/capability_test.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+				properties: nil,
+			},
+			want: nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := MakeCapapability(tt.args.c, tt.args.properties); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("MakeCapapability() = %v, want %v", got, tt.want)
+			}
```
Comment (meik99) [inline controllers/activegate/capability/capability_test.go:156]: Can be simplified
`assert.DeepEqual(t, tt.want, MakeCapapability(tt.args.c, tt.args.properties))`

File: controllers/activegate/capability/reconciler_test.go [hunk 8]
```diff
@@ -180,3 +196,57 @@ func TestSetReadinessProbePort(t *testing.T) {
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			r := &Reconciler{
+				Reconciler: tt.fields.Reconciler,
+				log:        tt.fields.log,
+				Capability: tt.fields.Capability,
+			}
+			if got := r.calculateStatefulSetName(); got != tt.want {
+				t.Errorf("Reconciler.calculateStatefulSetName() = %v, want %v", got, tt.want)
+			}
```
Comment (meik99) [inline controllers/activegate/capability/reconciler_test.go:249]: can be simplified
`assert.Equals(t, tt.want, r.calculateStatefulSetName())`

File: controllers/dynakube/dynakube_controller.go [hunk 9]
```diff
@@ -235,35 +220,50 @@ func (r *ReconcileDynaKube) ensureDeleted(obj client.Object) error {
...
+				svc := corev1.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      capability.BuildServiceName(rec.Instance.Name, cap.ModuleName),
+						Namespace: rec.Instance.Namespace,
+					},
+				}
+				if err := r.ensureDeleted(&svc); rec.Error(err) {
+					return false
+				}
+			}
```
Comment (meik99) [inline controllers/dynakube/dynakube_controller.go:263]: cool refactoring

File: controllers/activegate/capability/capability.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+			CapabilityName: "metrics_ingest",
+			Properties:     crProperties,
+			Configuration: Configuration{
+				SetDnsEntryPoint:     true,
+				SetReadinessPort:     true,
+				SetCommunicationPort: true,
+				CreateService:        true,
+			},
+		}
+	}
```
Comment (meik99) [inline controllers/activegate/capability/capability.go:73]: cool refactoring

File: controllers/dynakube/dynakube_controller_test.go [hunk 11]
```diff
@@ -127,7 +126,8 @@ func TestReconcileActiveGate_Reconcile(t *testing.T) {
 		assert.NotNil(t, result)
 
 		var statefulSet appsv1.StatefulSet
-		err = fakeClient.Get(context.TODO(), client.ObjectKey{Name: instance.Name + kubemon.StatefulSetSuffix, Namespace: testNamespace}, &statefulSet)
+		name := instance.Name + "-kubemon"
```
Comment (meik99) [inline controllers/dynakube/dynakube_controller_test.go:129]: please use constants for strings

**Review by meik99 — CHANGES_REQUESTED** [review]

File: controllers/activegate/capability/reconciler_test.go (hunk 8 above)
Comment (mjgrzybek) [reply controllers/activegate/capability/reconciler_test.go:249]: it's generated code with only input specified, I'd rather prefer not to change it

**Comment by mjgrzybek** [comment]
PR moved to: https://github.com/Dynatrace/dynatrace-operator/pull/135

File: controllers/activegate/capability/reconciler_test.go (hunk 8 above)
Comment (meik99) [reply controllers/activegate/capability/reconciler_test.go:249]: just out of curiosity, what tool did you use to generate these tests?

Comment (mjgrzybek) [reply controllers/activegate/capability/reconciler_test.go:249]: I'm using vscode + https://marketplace.visualstudio.com/items?itemName=golang.go
but AFAIR Intelij Goland used to generate the same output, so maybe some go backend tool provides this pattern?

**Pull request closed without merging**